### PR TITLE
[Inductor] Preserve StarDep/WeakDep fake deps in _compute_attrs

### PR DESCRIPTION
**Pull request by karthickai**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #178486

_compute_attrs calls extract_read_writes to re-derive the dependency set
for a scheduler node, but extract_read_writes only discovers real memory
accesses.  Manually-added StarDep and WeakDep ordering constraints were
silently dropped whenever _compute_attrs was called after initial
construction -- notably from recompute_size_and_body (CPU outer-loop
fusion path) and cancel_reduction_split.

The lost ordering deps allowed the scheduler to place a fused kernel
before its prerequisite buffer allocation, causing UnboundLocalError
in the generated backward code.

Apply the same fake-dep preservation pattern already used by
refresh_dependencies: save StarDep/WeakDep entries before re-extraction
and merge them back via with_read().

Fixes #175530

Authored with Claude Inductor Agent.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -1546,6 +1546,19 @@ def _compute_attrs(
...
+        # cannot be re-derived by extract_read_writes.  On the first call
+        # (from __init__) read_writes hasn't been set yet, so there is
+        # nothing to preserve.
+        fake_deps: OrderedSet[Dep] = OrderedSet()
+        if hasattr(self, "read_writes"):
+            fake_deps = OrderedSet(
+                dep
+                for dep in self.read_writes.reads
+                if isinstance(dep, (WeakDep, StarDep))
+            )
```
Comment (tianrengao) [inline torch/_inductor/scheduler.py]: nit: identical code as https://github.com/pytorch/pytorch/pull/178486/changes#diff-d2539c9c8dc6a3d7e457767a880612e96d3c85752a77ead49a9e4e00a3e4c3c7R1602-R1606, consider reuse code with a util or herper function

```
def _collect_fake_deps(self) -> OrderedSet[Dep]:
    if not hasattr(self, "read_writes"):
        return OrderedSet()
    return OrderedSet(
        dep for dep in self.read_writes.reads
        if isinstance(dep, (WeakDep, StarDep))
    )
```

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -1563,16 +1576,16 @@ def _compute_attrs(
...
-                dependencies.extract_read_writes(
-                    self._body, *self._sizes, normalize=should_normalize
-                )
+            new_rw = dependencies.extract_read_writes(
+                self._body, *self._sizes, normalize=should_normalize
             )
 
+        if fake_deps:
+            new_rw = new_rw.with_read(fake_deps)
+        self.set_read_writes(new_rw)
```
Comment (tianrengao) [inline torch/_inductor/scheduler.py:1614]: Do we need to call .rename similar to
https://github.com/pytorch/pytorch/pull/178486/changes#diff-d2539c9c8dc6a3d7e457767a880612e96d3c85752a77ead49a9e4e00a3e4c3c7L1601-L1602
```
        self.set_read_writes(
            dependencies.extract_read_writes(
                self._body, *self._sizes, normalize=normalize
            )
            .with_read(fake_deps)
            .rename(self.mutation_renames)
        )

```

if recompute_size_and_body is called after mutation renames are set up, will this lead to an issue due to inconsistent name dependency? It would be helpful to confirm it or add a comment if this never happens

File: test/inductor/test_cpu_repro.py [hunk 3]
```diff
@@ -6275,6 +6275,49 @@ def forward(x):
...
+
+        x1 = torch.randn(4, 128, 16, 16)
+        x2 = torch.randn(4, 128, 16, 16)
+
+        model = Model()
+        eager_model = model
+        compiled_model = torch.compile(copy.deepcopy(model))
+        compiled_loss = compiled_model(x1.clone(), x2.clone())
+        compiled_loss.backward()
+        eager_loss = eager_model(x1.clone(), x2.clone())
```
Comment (tianrengao) [inline test/inductor/test_cpu_repro.py:6328]: let's also assert close on gradients?

**Review by tianrengao — APPROVED** [review]
LGTM! A few comments

File: test/inductor/test_cpu_repro.py (hunk 3 above)
Comment (karthickai) [reply test/inductor/test_cpu_repro.py:6328]: Added assert close on grads

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (karthickai) [reply torch/_inductor/scheduler.py:1614]: _compute_attrs never had .rename(self.mutation_renames) before this PR either, But added the defensive rename anyway for consistency with refresh_dependencies.

File: torch/_inductor/scheduler.py (hunk 1 above)
Comment (karthickai) [reply torch/_inductor/scheduler.py]: refactored with helper function

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (tianrengao) [reply torch/_inductor/scheduler.py:1614]: Sounds good!

**Review by mlazos — APPROVED** [review]
